parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java

```diff
@@ -0,0 +1,316 @@
+      return ROWS_MIGHT_MATCH;
+    }
+
+    @Override
+    public <T> Boolean gt(BoundReference<T> ref, Literal<T> lit) {
+      // bloom filter is based on hash and cannot eliminate based on lt or ltEq or gt or gtEq
+      return ROWS_MIGHT_MATCH;
+    }
+
+    @Override
+    public <T> Boolean gtEq(BoundReference<T> ref, Literal<T> lit) {
+      // bloom filter is based on hash and cannot eliminate based on lt or ltEq or gt or gtEq
+      return ROWS_MIGHT_MATCH;
+    }
+
+    @Override
+    public <T> Boolean eq(BoundReference<T> ref, Literal<T> lit) {
+      int id = ref.fieldId();
+      if (!fieldsWithBloomFilter.contains(id)) { // no bloom filter
+        return ROWS_MIGHT_MATCH;
+      }
+      BloomFilter bloom = bloomById(id);
+      Type type = types.get(id);
+      T value = lit.value();
+      return shouldRead(parquetPrimitiveTypes.get(id), value, bloom, type);
+    }
+
+    @Override
+    public <T> Boolean notEq(BoundReference<T> ref, Literal<T> lit) {
+      // bloom filter is based on hash and cannot eliminate based on notEq
+      return ROWS_MIGHT_MATCH;
+    }
+
+    @Override
+    public <T> Boolean in(BoundReference<T> ref, Set<T> literalSet) {
+      int id = ref.fieldId();
+      if (!fieldsWithBloomFilter.contains(id)) { // no bloom filter
+        return ROWS_MIGHT_MATCH;
+      }
+      BloomFilter bloom = bloomById(id);
+      Type type = types.get(id);
+      for (T e : literalSet) {
+        if (shouldRead(parquetPrimitiveTypes.get(id), e, bloom, type)) {
+          return ROWS_MIGHT_MATCH;
+        }
+      }
+      return ROWS_CANNOT_MATCH;
+    }
+
+    @Override
+    public <T> Boolean notIn(BoundReference<T> ref, Set<T> literalSet) {
+      // bloom filter is based on hash and cannot eliminate based on notIn
+      return ROWS_MIGHT_MATCH;
+    }
+
+    @Override
+    public <T> Boolean startsWith(BoundReference<T> ref, Literal<T> lit) {
+      // bloom filter is based on hash and cannot eliminate based on startsWith
+      return ROWS_MIGHT_MATCH;
+    }
+
+    @Override
+    public <T> Boolean notStartsWith(BoundReference<T> ref, Literal<T> lit) {
+      // bloom filter is based on hash and cannot eliminate based on startsWith
+      return ROWS_MIGHT_MATCH;
+    }
+
+    private BloomFilter bloomById(int id) {
+      if (bloomCache.containsKey(id)) {
+        return bloomCache.get(id);
+      } else {
+        ColumnChunkMetaData columnChunkMetaData = columnMetaMap.get(id);
+        BloomFilter bloomFilter = bloomReader.readBloomFilter(columnChunkMetaData);
+        if (bloomFilter == null) {
+          throw new IllegalStateException("Failed to read required bloom filter for id: " + id);
+        } else {
+          bloomCache.put(id, bloomFilter);
+        }
+
+        return bloomFilter;
+      }
+    }
+
+    private <T> boolean shouldRead(PrimitiveType primitiveType, T value, BloomFilter bloom, Type type) {
+      long hashValue = 0;
+      switch (type.typeId()) {
```

Comment:
I think that this needs to use the Parquet type from the file, not the Iceberg type.

The Iceberg type comes from the read schema, which could be `long` when the file's type is `int` because of type promotion rules. We need to make sure that the hash function used matches the file's type because that was what produced the bloom filter. Hopefully, the same hash function is used for `long` and `int` in Parquet, but I'm not sure that is the case so the safest thing is to convert `value` to the Parquet type.

Reply:
I changed the code to `switch (primitiveType.getPrimitiveTypeName())`